src/Juvix/Compiler/Backend/Geb/Evaluator.hs

```diff
@@ -0,0 +1,377 @@
+
+type EvalEffects r = Members '[Reader Env, Error EvalError] r
+
+eval :: EvalEffects r => Morphism -> Sem r GebValue
+eval morph = do
+  -- env <- asks (^. envContext)
+  -- trace
+  --   ( "eval"
+  --       <> "\n"
+  --       <> "  arg:="
+  --       <> show morph
+  --       <> "\n"
+  --       <> "  env:="
+  --       <> show env
+  --       <> "\n"
+  --   ) $
+  case morph of
+    MorphismAbsurd x -> evalAbsurd x
+    MorphismApplication app -> evalApp app
+    MorphismBinop op -> evalBinop op
+    MorphismCase c -> evalCase c
+    MorphismFirst f -> evalFirst f
+    MorphismInteger i -> return $ GebValueMorphismInteger i
+    MorphismLambda l -> evalLambda l
+    MorphismLeft m -> GebValueMorphismLeft <$> eval m
+    MorphismPair p -> evalPair p
+    MorphismRight m -> GebValueMorphismRight <$> eval m
+    MorphismSecond s -> evalSecond s
+    MorphismUnit -> return GebValueMorphismUnit
+    MorphismVar x -> evalVar x
+
+evalVar :: EvalEffects r => Var -> Sem r GebValue
+evalVar var = do
+  ctx <- asks (^. envContext)
+  let val = Context.lookup (var ^. varIndex) ctx
+  -- trace ("varLookup := " <> show val) $
+  return val
+
+evalAbsurd :: EvalEffects r => Morphism -> Sem r GebValue
+evalAbsurd morph =
+  throw
+    EvalError
+      { _evalErrorMsg = "Absurd can not be evaluated.",
+        _evalErrorGebValue = Nothing,
+        _evalErrorGebExpression = Just morph
+      }
+
+evalPair :: EvalEffects r => Pair -> Sem r GebValue
+evalPair pair = do
+  left <- eval $ pair ^. pairLeft
+  right <- eval $ pair ^. pairRight
+  return $
+    GebValueMorphismPair $
+      ValueMorphismPair
+        { _valueMorphismPairLeft = left,
+          _valueMorphismPairRight = right
+        }
+
+evalFirst :: EvalEffects r => First -> Sem r GebValue
+evalFirst f = do
+  res <- eval $ f ^. firstValue
+  case res of
+    GebValueMorphismPair pair -> return $ pair ^. valueMorphismPairLeft
+    _ ->
+      throw
+        EvalError
+          { _evalErrorMsg = "First can only be applied to pairs.",
+            _evalErrorGebValue = Nothing,
+            _evalErrorGebExpression = Just (MorphismFirst f)
+          }
+
+evalSecond :: EvalEffects r => Second -> Sem r GebValue
+evalSecond s = do
+  res <- eval $ s ^. secondValue
+  case res of
+    GebValueMorphismPair pair -> return $ pair ^. valueMorphismPairRight
+    _ ->
+      throw
+        EvalError
+          { _evalErrorMsg = "Second can only be applied to pairs.",
+            _evalErrorGebValue = Just res,
+            _evalErrorGebExpression = Just (MorphismSecond s)
+          }
+
+evalApp :: EvalEffects r => Application -> Sem r GebValue
+evalApp app = do
+  evalStrategy <- asks (^. envEvaluatorOptions . evaluatorOptionsEvalStrategy)
+  let maybeForce :: GebValue -> GebValue
+      maybeForce = case evalStrategy of
+        CallByName -> id
+        CallByValue -> force
+  arg <- maybeForce <$> eval (app ^. applicationRight)
+  apply (app ^. applicationLeft) arg
```

Comment:
I don't think this gives you the evaluation strategy you want, because for `force` to be evaluated, `arg` needs to be evaluated, and it won't be until you force its evaluation, e.g., by pattern matching on it.

Besides, I don't think you need `deepseq` to achieve call-by-value. Call-by-value can be achieved just by forcing shallow evaluation of all function, constructor and primitive arguments. Because if you consistently evaluate all arguments, then anything contained in some GebValue will already be evaluated because it got forced earlier..

Reply:
And we don't really need those evaluation strategies. In STLC the result is always the same anyway. I doubt it would ever have any visible impact on performance in our usage.